srml/offences/src/lib.rs

```diff
@@ -75,57 +89,45 @@ decl_module! {
 	}
 }
 
-impl<T: Trait, O: Offence<T::IdentificationTuple>> ReportOffence<T::AccountId, T::IdentificationTuple, O>
-	for Module<T> where
+impl<T: Trait, O: Offence<T::IdentificationTuple>>
+	ReportOffence<T::AccountId, T::IdentificationTuple, O> for Module<T>
+where
 	T::IdentificationTuple: Clone,
 {
 	fn report_offence(reporters: Vec<T::AccountId>, offence: O) {
 		let offenders = offence.offenders();
 		let time_slot = offence.time_slot();
-		let session = offence.session_index();
 		let validator_set_count = offence.validator_set_count();
 
-		// Check if an offence is already reported for the offender authorities
-		// and otherwise stores that report.
-		let mut new_offenders = BTreeSet::new();
-		let all_offenders = {
-			// We have to return the modified list of offending authorities. If we were to use
-			// `mutate` we would have to clone the whole list in order to return it.
-			let mut offending_authorities = <OffenceReports<T>>::get(&O::ID, &(session, time_slot));
-
-			for offender in offenders {
-				// TODO [slashing] This prevents slashing for multiple reports of the same kind at the same slot,
-				// note however that we might do that in the future if the reports are not exactly the same (dups).
-				// De-duplication of reports is tricky though, we need a canonical form of the report
-				// (for instance babe equivocation can have headers swapped).
-				if !offending_authorities.iter().any(|details| details.offender == offender) {
-					new_offenders.insert(offender.clone());
-					offending_authorities.push(OffenceDetails {
-						offender,
-						reporters: reporters.clone().into_iter().collect(),
-					});
-				}
-			}
+		let mut cache = ReportIndexCache::load(&time_slot);
```

Comment:
Maybe better to store `new_offenders` in the `ReportIndexCache` and just return it from `collect_*` function?